core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -177,76 +177,43 @@ public long sizeInBytes() {
      */
-    public ImmutableTranslogReader immutableReader() throws TranslogException {
-        if (channelReference.tryIncRef()) {
-            synchronized (this) {
-                try {
-                    ensureOpen();
-                    outputStream.flush();
-                    ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channelReference, firstOperationOffset, getWrittenOffset(), operationCounter);
-                    channelReference.incRef(); // for new reader
-                    return reader;
-                } catch (Exception e) {
-                    throw new TranslogException(shardId, "exception while creating an immutable reader", e);
-                } finally {
-                    channelReference.decRef();
+    public synchronized ImmutableTranslogReader closeIntoReader() throws TranslogException, IOException {
+        if (closed.compareAndSet(false, true)) {
+            boolean success = false;
+            try {
+                sync();
+                ImmutableTranslogReader reader = new ImmutableTranslogReader(this.generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter);
+                success = true;
+                return reader;
+            } catch (Throwable e) {
+                closeWithTragicEvent(e);
+                throw new TranslogException(shardId, "exception while creating an immutable reader", e);
+            } finally {
+                if (success == false) {
+                    IOUtils.closeWhileHandlingException(channel);
                 }
             }
         } else {
-            throw new TranslogException(shardId, "can't increment channel [" + channelReference + "] ref count");
+            throw new AlreadyClosedException("translog [" + getGeneration() + "] is already closed", tragedy);
         }
     }
 
 
-    private long getWrittenOffset() throws IOException {
-        return channelReference.getChannel().position();
-    }
-
-    /**
-     * this class is used when one wants a reference to this file which exposes all recently written operation.
-     * as such it needs access to the internals of the current reader
-     */
-    final class InnerReader extends TranslogReader {
-
-        public InnerReader(long generation, long fistOperationOffset, ChannelReference channelReference) {
-            super(generation, channelReference, fistOperationOffset);
-        }
-
-        @Override
-        public long sizeInBytes() {
-            return TranslogWriter.this.sizeInBytes();
-        }
-
-        @Override
-        public int totalOperations() {
-            return TranslogWriter.this.totalOperations();
+    @Override
+    public synchronized Translog.Snapshot newSnapshot() {
+        try {
+            ensureOpen();
+            sync();
+            return super.newSnapshot();
+        } catch (IOException e) {
+            throw new TranslogException(shardId, "exception while a snapshot", e);
```

Comment:
"while a snapshot" -> "while creating a snapshot"?

I think `ensureOpen` can be moved outside of the try?

Can we maybe separate `sync()` into its own try, with a different message ("exception while syncing for snapshot") on exception?